packages/base-driver/lib/basedriver/capabilities.js

```diff
@@ -336,43 +335,74 @@ function processCapabilities(
- * will themselves be stripped. This is designed as an internal function, not one to operate on
- * user-constructed capabilities.
+ * capability and promoted it to the top level.
  *
- * @param {object} originalCaps - the capabilities to analyze and promote from 'options'
- * @return {object!} - the capabilities with 'options' promoted if necessary
+ * @template {Constraints} C
+ * @param {import('@appium/types').W3CCapabilities<C>} originalCaps
+ * @return {import('@appium/types').W3CCapabilities<C>} the capabilities with 'options' promoted if necessary
  */
 function promoteAppiumOptions(originalCaps) {
-  const appiumOptions = originalCaps[APPIUM_OPTS_CAP];
-  if (!appiumOptions) {
-    return originalCaps;
-  }
+  const promoteForObject = (obj) => {
+    const appiumOptions = obj[PREFIXED_APPIUM_OPTS_CAP];
+    if (!appiumOptions) {
+      return obj;
+    }
 
-  let caps = _.cloneDeep(originalCaps);
-  if (!_.isPlainObject(appiumOptions)) {
-    throw new errors.SessionNotCreatedError(`The ${APPIUM_OPTS_CAP} capability must be an object`);
-  }
+    if (!_.isPlainObject(appiumOptions)) {
+      throw new errors.SessionNotCreatedError(`The ${PREFIXED_APPIUM_OPTS_CAP} capability must be an object`);
+    }
+    if (_.isEmpty(appiumOptions)) {
+      return obj;
+    }
 
-  // first get rid of any prefixes inside appium:options
-  const strippedAppiumOptions = stripAppiumPrefixes(appiumOptions);
-  // warn if we are going to overwrite any keys on the base caps object
-  const overwrittenKeys = _.intersection(Object.keys(caps), Object.keys(strippedAppiumOptions));
-  if (overwrittenKeys.length > 0) {
-    log.warn(
-      `Found capabilities inside ${PREFIXED_APPIUM_OPTS_CAP} that will overwrite ` +
-        `capabilities at the top level: ${JSON.stringify(overwrittenKeys)}`
+    log.debug(
+      `Found ${PREFIXED_APPIUM_OPTS_CAP} capability present; will promote items inside to caps`
     );
-  }
 
-  // now just apply them to the main caps object
-  caps = {...caps, ...strippedAppiumOptions};
+    const shouldAddVendorPrefix = (/** @type {string} */ capName) => !capName.startsWith(APPIUM_VENDOR_PREFIX);
+    const verifyIfAcceptable = (capName) => {
+      if (!_.isString(capName)) {
+        throw new errors.SessionNotCreatedError(
+          `Capability names in ${PREFIXED_APPIUM_OPTS_CAP} must be strings. '${capName}' is unexpected`
+        );
+      }
+      if (isStandardCap(capName)) {
+        throw new errors.SessionNotCreatedError(
+          `${PREFIXED_APPIUM_OPTS_CAP} must only contain vendor-specific capabilties. '${capName}' is unexpected`
+        );
+      }
+      return capName;
+    };
+    const preprocessedOptions = _.toPairs(_.cloneDeep(appiumOptions))
+      .map(([name, value]) => [verifyIfAcceptable(name), value])
+      .map(([name, value]) => [shouldAddVendorPrefix(name) ? `${APPIUM_VENDOR_PREFIX}${name}` : name, value])
+      .reduce((acc, [name, value]) => {acc[name] = value; return acc;}, {});
```

Comment:
`appiumOptions` doesn't need to be cloned here because we do not mutate the value.

```suggestion
    const preprocessedOptions = _.mapKeys(
      _.mapKeys(appiumOptions, verifyIfAcceptable), 
      name => shouldAddVendorPrefix(name) ? `${APPIUM_VENDOR_PREFIX}${name}` : name, value])
    );
```

or if you wanna get real fancy

```suggestion
    const preprocessOptions = _.flow(
      _.mapKeys(_, verifyIfAcceptible),
      _.mapKeys(_, name => shouldAddVendorPrefix(name) ? `${APPIUM_VENDOR_PREFIX}${name}` : name, value])
    );
    const preprocessedOptions = preprocessOptions(appiumOptions);
```

Reply:
appiumOptions might have dictionary values. I need to make sure these are not going to be accidentally mutated in the returned value